src/district/server/smart_contracts.cljs

```diff
@@ -238,80 +238,95 @@
+                                                                      (log/debug "2) Callback get-past-events" {:contract contract
+                                                                                                                :event event
+                                                                                                                :from from
+                                                                                                                :to to})
+                                                                      (let [logs (->> events
+                                                                                      web3-helpers/js->cljkk
+                                                                                      (map (partial enrich-event-log contract contract-instance)))]
+                                                                        (async/put! ch-logs (or logs [(with-meta {:err error} {:error? true})]))
+                                                                        (async/close! ch-logs))))
+                                          ch-logs))
+                                      (async/merge)
+                                      (async/reduce into [])
+                                      (async/<!)
+                                      (sort-logs)
+                                      (async/>! ch))
+                                 (async/close! ch)))]
+    (async/pipeline-async 10 ch-ordered-logs chunk-for-all-events ch-from-blocks)
+
+    (go-loop [chunk-logs (async/<! ch-ordered-logs)]
+      (if chunk-logs
+        (do
+          (when (fn? callback)
+            (doseq [log chunk-logs]
+              (let [res (try
+                          (if-let [?error (:error? (meta log))]
+                            (callback ?error nil)
+                            (callback nil log))
+                          (catch js/Error e
+                            (when crash-on-event-fail?
+                              (log/error e "Server crash. Caused by event processing error with :crash-on-event-fail? true. Disable this flag to skip and continue.")
+                              (.exit js/process 1))))]
+                ;; if callback returns a promise or chan we block until it resolves
+                (cond
+                  (satisfies? cljs.core.async.impl.protocols/ReadPort res)
+                  (<! res)
+
+                  (async-helpers/promise? res)
+                  (<! (async-helpers/promise->chan res))))))
+
+          (log/debug "3) Processing chunks of blocks have finished. Call function on-chunk.")
+
+          (when (fn? on-chunk)
+            (on-chunk chunk-logs))
+
+          (recur (async/<! ch-ordered-logs)))
+        (do
+          (log/debug "4) Processing chunks of blocks have finished. Call function on-finish.")
+
+          (when (fn? on-finish)
```

Comment:
As an alternative of this and other conditionals based on if the function was given - you could have default value for these callbacks - a no-op (empty function). So you could just call it, thus simplifying. I prefer that style - it's more declarative (_always do this_) instead of checking and then doing.

But that's a bit matter of personal preference, so up to you.